Support for intermittent calls to update non-orographic gravity wave tendencies

ray-chew opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
<!--- One sentence to describe the purpose of this PR, refer to any linked issues:
#14 -- this will link to issue 14
Closes #2 -- this will automatically close issue 2 on PR merge
-->
Updated features and documentation of the non-orographic gravity module. The main features are:

1. Non-orographic gravity-wave tendencies are now diagnostic variables.
2. Non-orographic gravity-wave tendencies are only updated every `dt_nogw` for better computational efficiency, but the tendencies are still added after every dycore timestep.

Closes #3434 

## To-do
<!---  list of proposed tasks in the PR, move to "Content" on completion 
- Proposed task
-->

1. Add docstrings for non-orographic gravity-wave methods
2. Complete going through `gravity_wave.md`


## Content
<!---  specific tasks that are currently complete 
- Solution implemented
-->
1. ~~Added output diagnostics to the non-orographic config yaml files.~~ (Doesn't seem to be necessary)
3. Extended diagnostics to output (non-orographic) gravity-wave tendencies.
4. Fixed a typo in the example of `diagnostics.md`.
5. Slightly refactored `non_orographic_gravity_wave.jl`.
6. Changed all instances of `NonOrograph**y**GravityWave` type to `NonOrograph**ic**GravityWave`, to be consistent with the `OrographicGravityWave` type.
7. 1. Added `dt_nogw` as a default config. 
    3. The callback method `non_orographic_gravity_wave_compute_tendency!` now computes and updates the non-orographic tendencies every timestep size of `dt_nogw`.
    4. The tendencies are still called and updated every dycore timestep via `remanining_tendency.jl`.
8. Added an elaboration for $\epsilon$ in `gravity_wave.md`.
9. ~~Added `ClimaCoreTempestRemap` to `Project.toml` for orographic gravity wave unit tests.~~ (Not necessary if I use the correct project.)


<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [x] I have read and checked the items on the review checklist.

szy21 commented on config/model_configs/single_column_nonorographic_gravity_wave.yml:
```diff
@@ -4,5 +4,10 @@ t_end: "1500secs"
 config: "column"
 hyperdiff: false
 dt: "400secs"
+output_default_diagnostics: true
```
```suggestion
```
(This should be true by default)

szy21 commented on config/model_configs/single_column_nonorographic_gravity_wave.yml:
```diff
@@ -4,5 +4,10 @@ t_end: "1500secs"
+dt_nogw: "400secs"
+# diagnostics:
+#   - short_name: [ta, rhoa]
+#     period: 400secs
```
Add a new line

ray-chew replied:
```suggestion
#     period: 400secs

```

ray-chew replied:
```suggestion
#     period: 400secs

```

szy21 commented on config/model_configs/sphere_nonorographic_gravity_wave.yml:
```diff
@@ -1,4 +1,9 @@
-t_end: "1500secs"
+t_end: "12hours"
 dt: "400secs"
+deep_atmosphere: false
```
```suggestion
```

szy21 commented on docs/src/gravity_wave.md:
````diff
@@ -165,8 +165,9 @@ Similar to the base flux calculation but for the 3D fields, we computed ``N`` an
-^f V_{\tau}[k] = \max(\epsilon_0, - V[k] \cdot \frac{\tau}{|\tau|}).
+^f V_{\tau}[k] = \max(\epsilon_0, - V[k] \cdot \frac{\tau}{|\tau|}),
 ```
+where where ``\epsilon_0`` denotes a measure of floating-point precision.
````
```suggestion
where ``\epsilon_0`` denotes a measure of floating-point precision.
```

szy21 commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+        #     cache.non_orographic_gravity_wave.u_waveforcing,
+        #     axes(state.c),
+        # )
+        return copy(cache.non_orographic_gravity_wave.u_waveforcing)
```
```suggestion
        return copy(cache.non_orographic_gravity_wave.uforcing)
```
(same for other diagnostics)

szy21 commented on test/Project.toml:
You shouldn't need to change this file.

ray-chew commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+        #     cache.non_orographic_gravity_wave.v_waveforcing,
+        #     axes(state.c),
+        # )
+        return copy(cache.non_orographic_gravity_wave.v_waveforcing)
```
```suggestion
        return copy(cache.non_orographic_gravity_wave.vforcing)
```

ray-chew commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+        # )
+        return copy(cache.non_orographic_gravity_wave.v_waveforcing)
+    else
+        out .= cache.non_orographic_gravity_wave.v_waveforcing
```
```suggestion
        out .= cache.non_orographic_gravity_wave.vforcing
```

szy21 commented on config/default_configs/default_config.yml:
```diff
@@ -220,6 +220,9 @@ orographic_gravity_wave:
   value: false
+dt_nogw:
+  help: "Time between calling non-orographic gravity wave update"
+  value: "400secs"
```
```suggestion
  value: "1800secs"
```

dennisYatunin commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -368,49 +386,50 @@ function non_orographic_gravity_wave_forcing(
-    # loop over all wave lengths
-    for ink in 1:gw_nk
+    # Collect all required fields in a broadcasted object
+    input_u = Base.Broadcast.broadcasted(
+        tuple,
+        ᶜu_p1,
+        ᶜu_source,
+        ᶜbf_p1,
+        ᶜρ,
+        ᶜρ_p1,
+        ᶜρ_source,
+        ᶜz_p1,
+        ᶜz,
+        source_level,
+        gw_Bw,
+        gw_Bn,
+        gw_cw,
+        gw_cn,
+        gw_flag,
+        ᶜlevel,
+        gw_source_ampl,
+    )
```
```suggestion
    input_u = @. lazy(tuple(
        ᶜu_p1,
        ᶜu_source,
        ᶜbf_p1,
        ᶜρ,
        ᶜρ_p1,
        ᶜρ_source,
        ᶜz_p1,
        ᶜz,
        source_level,
        gw_Bw,
        gw_Bn,
        gw_cw,
        gw_cn,
        gw_flag,
        ᶜlevel,
        gw_source_ampl,
    ))
```
Maybe a little easier to read? Up to you.

ray-chew commented on test/Project.toml:
```diff
@@ -7,6 +7,7 @@ CairoMakie = "13f3f980-e62b-5c42-98c6-ff1f3baf88f0"
 ClimaAtmos = "b2c96348-7fb7-4fe0-8da9-78d88439e717"
 ClimaComms = "3a4d1b5c-c61d-41fd-a00a-5873ba7a1b0d"
 ClimaCore = "d414da3d-4745-48bb-8d80-42e94e092884"
+ClimaCoreTempestRemap = "d934ef94-cdd4-4710-83d6-720549644b70"
```
```suggestion
```

dennisYatunin commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -276,11 +269,36 @@ function non_orographic_gravity_wave_tendency!(
+    ::NonOrographicGravityWave,
+)
+
+    #unpack
```
```suggestion
```
Not a particularly helpful comment

ray-chew commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -368,49 +386,50 @@ function non_orographic_gravity_wave_forcing(
-            ᶜlevel,
-            gw_source_ampl,
-        )
+    input_v = Base.Broadcast.broadcasted(
+        tuple,
```
```suggestion
    input_v = @. lazy(tuple(
```

ray-chew commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -368,49 +386,50 @@ function non_orographic_gravity_wave_forcing(
+        gw_flag,
+        ᶜlevel,
+        gw_source_ampl,
+    )
```
```suggestion
    ))
```

szy21 commented:
@gemini-code-assist ignore

ray-chew commented on config/model_configs/sphere_nonorographic_gravity_wave.yml:
```diff
@@ -1,4 +1,9 @@
+dt_nogw: "400secs"
+diagnostics:
+  - short_name: [ta, thetaa, ua, va, wa, zg, utendnogw, vtendnogw]
+    period: 1hours
```
```suggestion
    period: 1hours
    
```

szy21 commented:
/gemini could you stop posting anything on this PR?

ray-chew commented:
Hi /gemini, please stop.

charleskawczynski commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+    time,
+    cache.atmos.non_orographic_gravity_wave,
+)
+compute_utendnogw!(_, _, _, _, non_orographic_gravity_wave::T) where {T} =
+    error_diagnostic_variable("utendnogw", non_orographic_gravity_wave)
```
```suggestion
compute_utendnogw!(_, _, _, _, non_orographic_gravity_wave) =
    error_diagnostic_variable("utendnogw", non_orographic_gravity_wave)
```

charleskawczynski commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+compute_utendnogw!(_, _, _, _, non_orographic_gravity_wave::T) where {T} =
+    error_diagnostic_variable("utendnogw", non_orographic_gravity_wave)
+
+function compute_utendnogw!(
+    out,
+    state,
+    cache,
+    time,
+    ::T,
+) where {T <: NonOrographicGravityWave}
```
```suggestion
function compute_utendnogw!(
    out,
    state,
    cache,
    time,
    ::NonOrographicGravityWave,
)
```

charleskawczynski commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+    time,
+    cache.atmos.non_orographic_gravity_wave,
+)
+compute_vtendnogw!(_, _, _, _, non_orographic_gravity_wave::T) where {T} =
+    error_diagnostic_variable("vtendnogw", non_orographic_gravity_wave)
```
```suggestion
compute_vtendnogw!(_, _, _, _, non_orographic_gravity_wave) =
    error_diagnostic_variable("vtendnogw", non_orographic_gravity_wave)
```

charleskawczynski commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+    state,
+    cache,
+    time,
+    ::T,
+) where {T <: NonOrographicGravityWave}
```
```suggestion
    ::NonOrographicGravityWave,
)
```

charleskawczynski commented on src/callbacks/get_callbacks.jl:
```diff
@@ -354,5 +354,21 @@ function get_callbacks(config, sim_info, atmos, params, Y, p)
     end
 
+    if atmos.non_orographic_gravity_wave isa NonOrographicGravityWave
+        dt_nogw =
+            dt isa ITime ? ITime(time_to_seconds(parsed_args["dt_nogw"])) :
+            FT(time_to_seconds(parsed_args["dt_nogw"]))
+        dt_nogw, _, _, _ = promote(dt_nogw, t_start, dt, sim_info.t_end)
```
I might prefer to define

```julia
time_type(::Type{ITime}, t) = ITime(t)
time_type(::Type{FT}, t) where {FT <: AbstractFloat} = FT(t)
```

and write this as 

```julia
        dt_nogw_str = parsed_args["dt_nogw"]::String
        dt_nogw_float = time_to_seconds(dt_nogw_str)::Float64
        dt_nogw = time_type(typeof(dt), dt_nogw_float)
        dt_nogw, _, _, _ = promote(dt_nogw, t_start, dt, sim_info.t_end)
```
but I'm wondering why the promote is needed, aren't all of the other time types using similar types? If not, then maybe we need a more principled initialization?

Sbozzolo replied:
There's no guarantee that `dt_nogw` and the others are represented with the same temporal basis. For example, `dt` could be 15 second (represented as 15 cycles of 1 second period) and `dt_nogw` could be 1 hour (represented as 1 cycle of 1 hour period). This promotion happens automatically when ITime is used, but sometimes is still useful to manually call `promote` because it allows one to assume that the types will be the same going forward (https://clima.github.io/ClimaUtilities.jl/dev/timemanager/#How-do-I-make-several-ITimes-have-the-same-type?). I am not sure if `promote` is strickly needed here, but if we are doing this for the other callbacks is likely because we found some edge case where it made things easier.

ray-chew replied:
Then I will keep the code as it is now.

charleskawczynski replied:
> There's no guarantee that dt_nogw and the others are represented with the same temporal basis.

I understand that, I'm not sure I read a suggestion or proposal in your message (@Sbozzolo), but that doesn't mean that this isn't a code smell. What I proposed does not change the behavior. Instead, I'm proposing defining functions that can be used here, and elsewhere, to reduce of complicated logic and the likelihood of inconsistencies.

Sbozzolo commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,86 @@
+    time,
+    ::T,
+) where {T <: NonOrographicGravityWave}
+    if isnothing(out)
+        # u_waveforcing is not a field:
+        # return Fields.array2field(
+        #     cache.non_orographic_gravity_wave.u_waveforcing,
+        #     axes(state.c),
+        # )
+        return copy(cache.non_orographic_gravity_wave.uforcing)
+    else
+        out .= cache.non_orographic_gravity_wave.u_waveforcing
+    end
```
We can start using the simpler system to write diagnostics
```suggestion
if pkgversion(ClimaDiagnostics) >= v"0.2.13"
    return cache.non_orographic_gravity_wave.u_waveforcing
else
    if isnothing(out)
        # u_waveforcing is not a field:
        # return Fields.array2field(
        #     cache.non_orographic_gravity_wave.u_waveforcing,
        #     axes(state.c),
        # )
        return copy(cache.non_orographic_gravity_wave.uforcing)
    else
        out .= cache.non_orographic_gravity_wave.u_waveforcing
    end
end
```

Sbozzolo replied:
Also, you should return Fields if you want to get a NetCDF output (because we can only interpolate Fields)

szy21 replied:
Actually it is a field (the comments are wrong, we should remove them).

ray-chew replied:
> Also, you should return Fields if you want to get a NetCDF output (because we can only interpolate Fields)

But it seems that `uforcing` and `vforcing` are already Fields, should I still explicitly return Field, like this?
```
        # return Fields.array2field(
        #     cache.non_orographic_gravity_wave.u_waveforcing,
        #     axes(state.c),
        # )
```

szy21 replied:
No, you don't need to.

ray-chew replied:
> Actually it is a field (the comments are wrong, we should remove them).

Yes, the comment was meant to say: if `uforcing` or `vforcing` is _not_ a field, then we have to explicitly convert them before outputting them. It was something I wrote myself as I was learning.

ray-chew commented on docs/src/diagnostics.md:
```diff
@@ -228,6 +224,11 @@ function compute_relative_humidity!(
+
+compute_relative_humidity!(out, state, cache, time) =
+    compute_relative_humidity!(out, state, cache, time, cache.atmos.moisture_model)
+compute_relative_humidity!(_, _, _, _, model::T) where {T} =
+    error_diagnostic_variable("relative_humidity", model)
```
Thank you, I am new to Julia and looked this up:

> In Julia, specifying model::T where {T} without constraining T is equivalent to using model::Any. This adds no value unless you intend to enforce specific type relationships or constraints.

I will this in mind in the future.

szy21 commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,90 @@
+    ::NonOrographicGravityWave,
+)
+if pkgversion(ClimaDiagnostics) >= v"0.2.13"
+    return cache.non_orographic_gravity_wave.u_waveforcing
```
```suggestion
    return cache.non_orographic_gravity_wave.uforcing
```
(I think you know, just commenting here in case you missed it.)

ray-chew replied:
Thanks, this has been fixed locally. :)

szy21 commented on src/diagnostics/gravitywave_diagnostics.jl:
```diff
@@ -0,0 +1,90 @@
+        # )
+        return copy(cache.non_orographic_gravity_wave.uforcing)
+    else
+        out .= cache.non_orographic_gravity_wave.u_waveforcing
```
```suggestion
        out .= cache.non_orographic_gravity_wave.uforcing
```

ray-chew replied:
This too has been fixed locally.

ray-chew commented:
# Results with different `dt_nogw` 
`orig` is the output from the run before moving the computation of NOGW tendencies to `callback`.

![image](https://github.com/user-attachments/assets/3d7cb14d-4560-4ce7-b23d-5a366eca4286)
![image](https://github.com/user-attachments/assets/75b26763-8500-4009-bc09-324002daea1f)

szy21 commented on config/model_configs/single_column_nonorographic_gravity_wave.yml:
```diff
@@ -6,3 +6,7 @@ hyperdiff: false
 disable_surface_flux_tendency: true
 non_orographic_gravity_wave: true
+dt_nogw: "400secs"
+# diagnostics:
+#   - short_name: [ta, rhoa]
+#     period: 400secs
```
remove this?

ray-chew replied:
Done. Will be more careful with comments next time!

szy21 commented on config/model_configs/sphere_nonorographic_gravity_wave.yml:
```diff
@@ -1,4 +1,9 @@
+diagnostics:
+  - short_name: [ta, thetaa, ua, va, wa, zg, utendnogw, vtendnogw]
+    period: 1hours
+    
```
somehow github is not happy with the format, maybe you have some spaces on line 9? (if you can't figure it out, don't worry about it).

ray-chew replied:
Removed the spaces.

szy21 commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -276,11 +269,35 @@ function non_orographic_gravity_wave_tendency!(
         p,
     )
 
+    # # update cache after computation
+    # # do I need to do this? Since these attributes were
+    # # unpacked via reference?
+    # p.non_orographic_gravity_wave.uforcing .= uforcing
+    # p.non_orographic_gravity_wave.vforcing .= vforcing
```
Could you remove these comments?

ray-chew replied:
Done!

szy21 reviewed (approved):
Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.

ray-chew commented:
> Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.

I won't squash for now, but will do so in the future. I lumped a few unrelated things together in this PR, so keeping the commits separate makes sense to me. But I will make sure to submit strictly one-topic PRs next time.

Sbozzolo commented:
> > Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.
> 
> I won't squash for now, but will do so in the future. I lumped a few unrelated things together in this PR, so keeping the commits separate makes sense to me. But I will make sure to submit strictly one-topic PRs next time.

It's fine (and preferable) to have separate commits when the changes are logically distinct.

Sbozzolo commented on src/callbacks/callbacks.jl:
```diff
@@ -330,6 +330,16 @@ NVTX.@annotate function rrtmgp_model_callback!(integrator)
     return nothing
 end
 
+
+
+NVTX.@annotate function nogw_model_callback!(integrator)
+    Y = integrator.u
+    p = integrator.p
+
+    non_orographic_gravity_wave_compute_tendency!(Y, p)
+end
+
+
```
```suggestion

NVTX.@annotate function nogw_model_callback!(integrator)
    Y = integrator.u
    p = integrator.p

    non_orographic_gravity_wave_compute_tendency!(Y, p)
    return nothing
end

```

ray-chew replied:
Done.

Sbozzolo commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -276,11 +269,29 @@ function non_orographic_gravity_wave_tendency!(
 
 end
 
+
+
+
+
```
```suggestion
```

ray-chew replied:
Done.

ray-chew commented:
> > > Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.
> > 
> > 
> > I won't squash for now, but will do so in the future. I lumped a few unrelated things together in this PR, so keeping the commits separate makes sense to me. But I will make sure to submit strictly one-topic PRs next time.
> 
> It's fine (and preferable) to have separate commits when the changes are logically distinct.

Yep, I already squashed all intermediate commits, only the main ones are left (and commits for review comments).

Sbozzolo commented on src/callbacks/get_callbacks.jl:
```diff
@@ -354,5 +354,21 @@ function get_callbacks(config, sim_info, atmos, params, Y, p)
+            @warn "This simulation will not be reproducible when restarted"
+        end
+
+        callbacks = (callbacks..., call_every_dt(nogw_model_callback!, dt_nogw))
```
Note that this callback modifies the state, so the order of callback matters.

I don't know whether this is important or not.

szy21 replied:
I think this callback only modifies the tendencies, not the state?

Sbozzolo replied:
Oh, I was misled by how the function is called: 
```
non_orographic_gravity_wave_compute_tendency!(Y, p)
```

But it makes sense, we are already doing this in other places!

Sbozzolo reviewed (commented):
If would be great if you took the opportunity to improve the documentation for this feature.

There's a page about gravity waves, is it reflective of what's implemented? How do users enable gravity waves in our code? 

Also, please, update the NEWS file.

Sbozzolo commented:
> > > > Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.
> > > 
> > > 
> > > I won't squash for now, but will do so in the future. I lumped a few unrelated things together in this PR, so keeping the commits separate makes sense to me. But I will make sure to submit strictly one-topic PRs next time.
> > 
> > 
> > It's fine (and preferable) to have separate commits when the changes are logically distinct.
> 
> Yep, I have already squashed all intermediate commits, only the main ones are left (and commits for review comments).

Also, commit titles are typically in the imperative form and capitalized.

Here's some commonly adopted conventions: https://cbea.ms/git-commit/ (scroll down to the 7 rules)

ray-chew commented:
> > > > > Looks good, thanks! It's up to you whether you want to squash the commits. I don't have a strong preference. Maybe @charleskawczynski or @Sbozzolo has some thoughts.
> > > > 
> > > > 
> > > > I won't squash for now, but will do so in the future. I lumped a few unrelated things together in this PR, so keeping the commits separate makes sense to me. But I will make sure to submit strictly one-topic PRs next time.
> > > 
> > > 
> > > It's fine (and preferable) to have separate commits when the changes are logically distinct.
> > 
> > 
> > Yep, I have already squashed all intermediate commits, only the main ones are left (and commits for review comments).
> 
> Also, commit titles are typically in the imperative form and capitalized.
> 
> Here's some commonly adopted conventions: https://cbea.ms/git-commit/ (scroll down to the 7 rules)

Thank you, I will keep this in mind in the future.

I have updated my commit messages and further squashed my commits accordingly.

Sbozzolo commented on config/model_configs/single_column_nonorographic_gravity_wave.yml:
```diff
@@ -6,3 +6,5 @@ hyperdiff: false
 disable_surface_flux_tendency: true
 non_orographic_gravity_wave: true
+dt_nogw: "400secs"
+
```
```suggestion
```

Sbozzolo commented on src/diagnostics/diagnostic.jl:
```diff
@@ -98,7 +98,9 @@ include("core_diagnostics.jl")
+include("gravitywave_diagnostics.jl")
 include("conservation_diagnostics.jl")
 
+
```
```suggestion
```

Sbozzolo commented on src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl:
```diff
@@ -276,11 +269,29 @@ function non_orographic_gravity_wave_tendency!(
 
 end
 
+
+
+
+
```
```suggestion
```

Sbozzolo replied:
Please, remove all these extra spaces

ray-chew replied:
> Please, remove all these extra spaces

Thanks, but is there a formatting I should follow? I thought the formatter would catch these, but it didn't, and I noticed that sometimes there are two spaces between functions and sometimes one.

Please let me know what I should do. For now, I keep one line break for the modifications I made.

ray-chew replied:
> No, the formatter does not catch these things.
> 
> I don't think have a consistent style with respect to the number of lines between functions: one or two is fine, but four seems excessive

Okay, got it! Thanks.

Sbozzolo commented on config/default_configs/default_config.yml:
```diff
@@ -220,6 +220,9 @@ orographic_gravity_wave:
   help: "Apply parameterization for convective gravity wave forcing on horizontal mean flow [`false` (default), `true`]"
   value: false
+dt_nogw:
+  help: "Time between calling non-orographic gravity wave update"
```
```suggestion
  help: "Time between calling non-orographic gravity wave update. Ignored if `non_orographic_gravity_wave` is not set to `true`."
```

If there's anything else you want to add here, please go ahead. Best to have more information than less (don't emulate the other `help` sections in this file).

ray-chew replied:
Thanks for pointing this out. As mentioned during the Monday meeting, I will go through the documentations thoroughly after I am done with the orographic GW module. This is because things might change. E.g., we may have just one dt_gw in the end, for example.

ray-chew commented:
So I did the following:
```
git fetch upstream main
git rebase upstream/main
git push origin main --force-with-lease
```
This gave the error: "This branch is out-of-date with the base branch".

Somehow
```
git fetch --all
```
fixed the issue. Strange.

ray-chew commented:
Is there something I should do, or should I merge this PR, or should someone else do it?

szy21 commented:
It seems to have conflicts with the main branch. After resolving them, feel free to click `Merge when ready` at the bottom.

charleskawczynski reviewed (approved):
Overall I think this PR looks fine. I did leave a comment about some complicated [promotion/conversion logic that seems to be getting duplicated](https://github.com/CliMA/ClimaAtmos.jl/pull/3761/files#r2024994927), but I won't die on that hill defending my suggestion.

I personally prefer to lean towards writing small and simple type-stable functions that can be unit tested (to prevent death by a thousand cuts of performance or latency), which is what I proposed. Perhaps others feel differently.

ray-chew commented:
> It seems to have conflicts with the main branch. After resolving them, feel free to click `Merge when ready` at the bottom.

Thank you, I resolved the conflicts and clicked `Merge when ready`.

> Overall I think this PR looks fine. I did leave a comment about some complicated [promotion/conversion logic that seems to be getting duplicated](https://github.com/CliMA/ClimaAtmos.jl/pull/3761/files#r2024994927), but I won't die on that hill defending my suggestion.
> 
> I personally prefer to lean towards writing small and simple type-stable functions that can be unit tested (to prevent death by a thousand cuts of performance or latency), which is what I proposed. Perhaps others feel differently.

@charleskawczynski I realised I didn't quite understand your comment. I will reach out to you via Slack.

szy21 commented:
All the ci breaks and the buildkite doesn't start - I wonder if it is because this is a PR from a fork rather than a branch? @charleskawczynski @Sbozzolo should know more.

Sbozzolo commented:
@ray-chew do you have a buildkite account? I don't see your user

Sbozzolo commented:
As for GitHub actions, I think this was a problem in the codecov action, but it was fixed upstream. Re-running wll fix the problem.

ray-chew commented:
> @ray-chew do you have a buildkite account? I don't see your user

Nop, do I need one? I haven't come across this in the documentation, and I haven't used buildkite before.

Sbozzolo commented:
> > @ray-chew do you have a buildkite account? I don't see your user
> 
> Nop, do I need one? I haven't come across this in the documentation, and I haven't used buildkite before.
> 
> Edit: So I tried creating an account on buildkite, but the only option available to me is to start a 30-day trial, should I do this @Sbozzolo ?

You should received an invite to your caltech account.

ray-chew commented:
Do the codecovs have to pass?

Sbozzolo commented:
> How can I make codecov pass? Does it use the runs in the buildkite pipeline?

It's failing because your PR increases the percentage of untested lines (but it's fine). Ideally you'd want to test all your lines, but it's hard to do with climaatmos in Github Actions.

I don't know why buildkite is not starting (maybe because you got an account after you opened the PR). Could you try running this PR manually in buildkite? We can manually merge it when we see it's green

ray-chew commented:
> > How can I make codecov pass? Does it use the runs in the buildkite pipeline?
> 
> It's failing because your PR increases the percentage of untested lines (but it's fine). Ideally you'd want to test all your lines, but it's hard to do with climaatmos in Github Actions.

Ok. I don't know how codecov works. But if it starts a bunch of run scripts to test code coverage, then I can add a test case that uses the features I implemented in this PR.

> I don't know why buildkite is not starting (maybe because you got an account after you opened the PR). Could you try running this PR manually in buildkite? We can manually merge it when we see it's green

I will try running the CI manually.

charleskawczynski commented:
> I don't know why buildkite is not starting

I think it's because of what @szy21 said, it's a PR from a fork.

ray-chew commented:
> > I don't know why buildkite is not starting
> 
> I think it's because of what @szy21 said, it's a PR from a fork.

Yes, I think I found out enough to come to this conclusion, too. I prefer to work with my own fork so that I can create a mess there. But I will push to a branch here in the future and open the PR from the branch. Will close this PR and open a new one, since I can't change the source branch here.

ray-chew commented:
I suppose that one could enable this last option on Buildkite, too:
![image](https://github.com/user-attachments/assets/349f59bd-7df7-455e-9f06-5c3d22b0a88a)

But I will leave the decision to you all.

charleskawczynski commented:
> I suppose that one could enable this last option on Buildkite, too...

> But I will leave the decision to you all, @charleskawczynski @Sbozzolo @szy21 .

I think that that would fix the immediate issue, but I think that fixing it that way (without any other measures) could be a security risk. I recommend reaching out to Julia folks, and see how they handle this. I think a more secure approach would be to allow maintainers, or a smaller group of CliMA maintainers, to approve / allow fork builds to begin. That way, there is at least a way to enforce responsibility / accountability.

ray-chew commented:
> > I suppose that one could enable this last option on Buildkite, too...
> 
> > But I will leave the decision to you all, @charleskawczynski @Sbozzolo @szy21 .
> 
> I think that that would fix the immediate issue, but I think that fixing it that way (without any other measures) could be a security risk. I recommend reaching out to Julia folks, and see how they handle this. I think a more secure approach would be to allow maintainers, or a smaller group of CliMA maintainers, to approve / allow fork builds to begin. That way, there is at least a way to enforce responsibility / accountability.

I would be interested in finding a solution, too. We could find out how the people at `Oceananigans.jl` do it. They seem to have a fork-based approach to contributions, as opposed to a branch-one like ClimaAtmos. And they are also using Buildkite.

szy21 commented:
superseded by #3775
